IBD

danking opened the pull request:
resolves #180

An implementation of the algorithm for IBD described in the plink
paper.

danking commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  def similarityMatrix(gs: Iterable[Genotype]): Array[Array[IBS]] = {
+//      gs.map[Array[IBS], Array[Array[IBS]]](l => gs.map[IBS, Array[IBS]](r => genotypeSimilarity(l, r)))
```
I got weird type errors with this line, not sure what's wrong.

danking commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,165 @@
+    val a10 = 4 * p * p * p * q * ((x - 1) / x * (x - 2) / x * (Na / (Na - 1)) * (Na / (Na - 2)) * (Na / (Na - 3))) + 4 * p * q * q * q * ((y - 1) / y * (y - 2) / y * (Na / (Na - 1)) * (Na / (Na - 2)) * (Na / (Na - 3)))
+    val a20 = q * q * q * q * ((y - 1) / y * (y - 2) / y * (y - 3) / y * (Na / (Na - 1)) * (Na / (Na - 2)) * (Na / (Na - 3))) + p * p * p * p * ((x - 1) / x * (x - 2) / x * (x - 3) / x * (Na / (Na - 1)) * (Na / (Na - 2)) * (Na / (Na - 3))) + 4 * p * p * q * q * ((x - 1) / x * (y - 1) / y * (Na / (Na - 1)) * (Na / (Na - 2)) * (Na / (Na - 3)))
+    val a11 = 2 * p * p * q * ((x - 1) / x * Na / (Na - 1) * Na / (Na - 2)) + 2 * p * q * q * ((y - 1) / y * Na / (Na - 1) * Na / (Na - 2))
+    val a21 = p * p * p * ((x - 1) / x * (x - 2) / x * Na / (Na - 1) * Na / (Na - 2)) + q * q * q * ((y - 1) / y * (y - 2) / y * Na / (Na - 1) * Na / (Na - 2)) + p * p * q * ((x - 1) / x * Na / (Na - 1) * Na / (Na - 2)) + p * q * q * ((y - 1) / y * Na / (Na - 1) * Na / (Na - 2))
```
I literally copy-pasted these from the source code of plink 1.07

danking commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,165 @@
+    val x = gs.iterator.filter(_.gt.isDefined).map(_.gt.get).map(countRefs).sum.toDouble
+    val y = Na - x
+    val p = x / Na
+    val q = y / Na
```
In Plink 1.07, each locus has a `freq` attribute which is used for `p`.

danking commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,165 @@
+  def hasNaNs: Boolean = Array(E00, E10, E20, E11, E21).exists(_.isNaN)
+
+  def normalized: IBSExpectations =
+    IBSExpectations(E00 / nonNaNCount, E10 / nonNaNCount, E20 / nonNaNCount, E11 / nonNaNCount, E21 / nonNaNCount, E22, this.nonNaNCount)
```
NB: I don't normalize E22. In Plink 1.07, it's always equal to 1, AFIACT.

danking commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,165 @@
+    if (this.hasNaNs)
+      that
+    else if (that.hasNaNs)
+      this
```
These two cases are how I avoid counting IBS expectations with NaNs. In Plink, there's an if statement that achieves the same effect.

danking commented:
@cseed:

1. We talked about finding a better way to implement this so that it's faster. I have thus far focused only on the tests and the core algorithm. The `Command` and associated docs are not yet written.
2. I will extract `assertSameElements` and the fuzzy equality stuff into separate PRs before submitting this for review.

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/IBDCommand.scala:
```diff
@@ -0,0 +1,58 @@
+    var computeMafExpr: String = _
+
+    @Args4jOption(required = false, name = "--unbounded",
+      usage = "Allows the estimations for Z0, Z1, Z2, and PI_HAT to take on biologically-nonsense values.")
```
Clarify here?  Not sure what exactly this means (I assume negative)

danking replied:
Z0, Z1, and Z2 are supposed to sum to one. IIRC, I lifted this line from the PLINK paper. I'll add something.

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/IBDCommand.scala:
```diff
@@ -0,0 +1,58 @@
+    { (v: Variant, va: Annotation) =>
+      mafEc.setAll(v, va)
+      val maf = computeMafThunk()
+        .getOrElse(throw new RuntimeException(s"The minor allele frequency expression evaluated to NA on variant $v."))
```
why not `fatal`?

tpoterba commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,233 @@
+}
+
+case class IBDInfo(Z0: Double, Z1: Double, Z2: Double, PI_HAT: Double) {
+  def pointwiseMinus(that: IBDInfo): (Double, Double, Double, Double) =
```
something something 4-tuple

tpoterba commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,233 @@
+}
+
+case class ExtendedIBDInfo(ibd: IBDInfo, ibs0: Int, ibs1: Int, ibs2: Int) {
+  def pointwiseMinus(that: ExtendedIBDInfo): ((Double, Double, Double, Double), Double, Double, Double) =
```
same

tpoterba commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  def ibsForGenotypes(gs: Iterable[Genotype], maybeMaf: Option[Double]): IBSExpectations = {
+    val Na = gs.iterator.count(_.gt.isDefined) * 2.0
```
no need for iterator here

tpoterba commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,233 @@
+      val q = maf
+      val x = Na * p
+      val y = Na * q
+      (x, y, p ,q)
```
misformatted.  Also, 4-tuple but that might be OK here

tpoterba commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    def estimateFrequenciesFromSample = {
+      val x = gs.iterator.filter(_.gt.isDefined).map(_.gt.get).map(countRefs).sum.toDouble
```
slaps for `.get`.
Can be written as
```scala
gs.fold(0d) { case (sum, g) => sum + g.gt.map(countRefs).getOrElse(0) }
```

tpoterba commented on src/main/scala/org/broadinstitute/hail/methods/IBD.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  def makeTriangleNoDiagonal[T, U : ClassTag](ts: IndexedSeq[T], f: (T, T) => U): NoDiagonalTriangleMatrix[U] = {
+    val arr = new Array[Array[U]](ts.size - 1)
```
I think you can write this stuff with more scala-like syntax with:

```scala
val arr = Array.tabulate(ts.size - 1) { i => 
  ...
```

danking commented:
@tpoterba, I addressed your comments, thanks! all good points

@cseed, I think this is what you had in mind. I basically repartition the RDD by sample, call `cartesian` and then compute each position in the matrix. Take a look at lines [176-185](https://github.com/danking/hail/blob/issue-180/src/main/scala/org/broadinstitute/hail/methods/IBD.scala#L176-L185)

[I write the RDD containing the matrix to a file](https://github.com/danking/hail/blob/issue-180/src/main/scala/org/broadinstitute/hail/driver/IBDCommand.scala#L60-L65) before returning from the Command without changing the state. Seems OK? Baring larger changes to the structure of Hail (distribution along sample axis), I don't see a better option.

danking commented:
I think this might still use way to much memory.

danking commented:
```
dking@wmb16-359 # export JAVA_OPTS="-Xmx8g"                                                                  
dking@wmb16-359 # time ../hail/build/install/hail/bin/hail importvcf -n 256 profile.vcf.bgz grm -f rel -o bar
hail: info: running: importvcf -n 256 profile.vcf.bgz
[Stage 0:=====================================================> (207 + 4) / 214]hail: info: Coerced sorted dataset
hail: info: running: grm -f rel -o bar
[Stage 9:====================================================>      (8 + 1) / 9]hail: info: while importing:
    file:/Users/dking/projects/hail-data/profile.vcf.bgz  import clean
hail: info: timing:
  importvcf: 10.524s
  grm: 5m59.8s
../hail/build/install/hail/bin/hail importvcf -n 256 profile.vcf.bgz grm -f    1134.91s user 39.57s system 312% cpu 6:15.31 total
dking@wmb16-359 # (cd ../hail && gradle installDist) && (cd ../hail-data && rm -rf foo && time ../hail/build/install/hail/bin/hail importvcf profile.vcf.bgz splitmulti ibd -o 'foo') 
:compileJava UP-TO-DATE
:compileScala
:processResources UP-TO-DATE
:classes
:jar
:startScripts
:installDist

BUILD SUCCESSFUL

Total time: 18.115 secs
hail: info: running: importvcf profile.vcf.bgz
[Stage 0:==========================================>                (5 + 2) / 7]hail: info: Coerced sorted dataset
hail: info: running: splitmulti
hail: info: running: ibd -o foo
[Stage 2:==========================================>                (5 + 2) / 7]chunkSize 161, nSampleChunks 16, nSamples 2535
[Stage 12:=================================================>        (6 + 1) / 7]hail: info: while importing:
    file:/Users/dking/projects/hail-data/profile.vcf.bgz  import clean
hail: info: timing:
  importvcf: 6.277s
  splitmulti: 44.873ms
  ibd: 8m18.2s
../hail/build/install/hail/bin/hail importvcf profile.vcf.bgz splitmulti ibd   1639.38s user 9.14s system 324% cpu 8:28.71 total
```
on my MBP.

danking commented:
On my MBP with `JAVA_OPTS=-Xmx8g`,

 - `time hail importvcf -n 256 profile.vcf.bgz grm -f rel -o bar`
```
...
hail: info: timing:
  importvcf: 10.524s
  grm: 5m59.8s
hail importvcf -n 256 profile.vcf.bgz grm -f    1134.91s user 39.57s system 312% cpu 6:15.31 total
```
 - `time hail importvcf -n 256 profile.vcf.bgz splitmulti ibd -o 'foo'`
```
...
hail: info: timing:
  importvcf: 8.345s
  splitmulti: 23.195ms
  ibd: 8m32.4s
hail importvcf -n 256 profile.vcf.bgz  ibd -o   1761.34s user 24.76s system 339% cpu 8:45.80 total
```

cseed commented:
I agree, that seems slow for profile.  How long does PLINK take?

danking commented:
PLINK takes 16 seconds.

danking commented:
@cseed on the cray on the danish blood spot:

```
hail: info: running: read -i /user/satterst/DBS_v3/DBS_v3_split_vep.vds                                                                      
[Stage 1:======================================================>(352 + 1) / 353]hail: info: running: filtervariants intervals --keep -i file:///mnt/lustre/cseed/purcell5k.interval_list                                                                                                  
hail: info: running: ibd -o file:///mnt/lustre/dking/foo                                                                                     
[Stage 3:===============================================>   (2981 + 195) / 3176]                                                             
[Stage 8:====================================================>(3176 + 1) / 3177]hail: info: timing:                                          
  read: 8.071s                                                                                                                               
  filtervariants intervals: 186.551ms                                                                                                        
  ibd: 49m53.2s                                                                                                                              
```

danking commented:
Anecdotally, I saw it stuck on the last 4 partitions of Stage 7 for quite a while.

danking commented:
Here's what it looks like now. Note that I've already converted to a `vds` this time.

```
dking@wmb16-359 # rm -rf foo && time ../hail/build/install/hail/bin/hail read -i profile.vds ibd -o 'foo' 
hail: info: running: read -i profile.vds
[Stage 1:>                                                          (0 + 0) / 4]SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[Stage 1:============================================>              (3 + 1) / 4]hail: info: running: ibd -o foo
[Stage 8:=====================================================> (210 + 4) / 214]hail: info: timing:
  read: 3.047s
  ibd: 4m35.1s
../hail/build/install/hail/bin/hail read -i profile.vds ibd -o 'foo'  924.50s user 16.11s system 333% cpu 4:42.04 total
```

danking commented:
@tpoterba @cseed 

If y'all can take a look at the docs, tests, and implementation, I want to merge this. I included a log of running on `profile.vcf.bgz` (which has 2500 samples) below, total time is about 3.5 minutes. I expect it to scale roughly like `O(nSamples^2)`. For Kyle's use case this performance is acceptable.

Further performance, model, and usability improvements will be separate PRs.

```
dking@wmb16-359 # hail read -i profile.vds ibd -o foo
hail: info: running: read -i profile.vds
[Stage 0:==============>                                            (1 + 3) / 4]SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[Stage 1:============================================>              (3 + 1) / 4]hail: info: running: ibd -o foo
[Stage 8:==================================================>    (197 + 4) / 214]hail: info: timing:
  read: 3.523s
  ibd: 3m33.8s

```

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/IBDCommand.scala:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  override protected def run(state: State, options: Options): State = {
+    val computeMaf = if (newOptions.computeMafExpr == null)
```
This could be `Option(newOptions.computeMafExpr.map(expr => generateComputeMaf(state.vds, expr))`

tpoterba commented on src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -294,6 +294,10 @@ object VSMSubgen {
     vGen = Variant.gen,
     tGen = Genotype.genExtreme)
 
+  val plinkSafeSingleAlt = random.copy(
```
I think this should be renamed to `plinkSafeBiallelic`, and should gen.map the vds to have `wasSplit = true` at the end

danking commented:
@tpoterba addressed
